### PR TITLE
sync P4 dpdk target with Open-source P4 DPDK target

### PR DESCRIPTION
**Pull request by riyazahm22**
sync P4 dpdk target with Open-source P4 DPDK target

### CODE REVIEW

**Comment by riyazahm22** [comment]
PNA Test Example
[root@localhost bin]# ./bf_switchd --install-dir $SDE_INSTALL --conf-file $SDE_INSTALL/share/examples/simple_lpm/simple_lpm.conf --init-mode=cold --status-port 7777
Install dir: /home/rahmad/p4-dpdk-open/install (0x9d0b80)
bf_switchd: system services initialized
bf_switchd: loading conf_file /home/rahmad/p4-dpdk-open/install/share/examples/simple_lpm/simple_lpm.conf...
bf_switchd: processing device configuration...
Configuration for dev_id 0
Family : dpdk
bf_switchd: processing P4 configuration...
P4 profile for dev_id 0
P4 EAL args dummy -n 4 -c 3
Debug CLI disable
num mempool objs 1
mempool_name: MEMPOOL0
buffer_size: 2304
pool_size: 1024
cache_size: 256
numa_node: 0
num P4 programs 1
p4_name: sample_lpm
bfrt_config: /home/rahmad/p4-dpdk-open/install/share/examples/simple_lpm/bf-rt.json
port_config: /home/rahmad/p4-dpdk-open/install/share/examples/simple_lpm/port_config.json
p4_pipeline_name: pipe
core_id: 1
numa_node: 0
context: /home/rahmad/p4-dpdk-open/install/share/examples/simple_lpm/pipe/context.json
config: /home/rahmad/p4-dpdk-open/install/share/examples/simple_lpm/pipe/simple_lpm.spec
Pipes in scope [0 1 2 3 ]
Mirror Config
n_slots: 4
n_sessions: 64
fast_clone: 0
diag:
accton diag:
non_default_port_ppgs: 0
SAI default initialize: 1
Operational mode set to DPDK MODEL
Initialized the device types using PAL handler registration
STUB:port_mgr_platform_init DPDK
bf_switchd: drivers initialized
EAL: Detected CPU lcores: 16
EAL: Detected NUMA nodes: 1
EAL: Detected shared linkage of DPDK
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'PA'
EAL: Probe PCI driver: net_virtio (1af4:1000) device: 0000:00:03.0 (socket 0)
eth_virtio_pci_init(): Failed to init PCI device
EAL: Requested device 0000:00:03.0 cannot be used
EAL: Probe PCI driver: net_virtio (1af4:1000) device: 0000:00:04.0 (socket 0)
eth_virtio_pci_init(): Failed to init PCI device
EAL: Requested device 0000:00:04.0 cannot be used
TELEMETRY: No legacy callbacks, legacy socket not created

bf_switchd: dev_id 0 initialized

bf_switchd: initialized 1 devices
bf_switchd_switchsai_lib_init
starting SAI
bf_switchd: spawning cli server thread
bf_switchd: spawning driver shell
bf_switchd: server started - listening on port 9999

    ********************************************
    *      WARNING: Authorised Access Only     *
    ********************************************
bfshell> bfrt_python
cwd : /home/rahmad/p4-dpdk-open/install/bin

Devices found : [0]
We've found 1 p4 programs for device 0:
sample_lpm
Creating tree for dev 0 and program sample_lpm

/home/rahmad/p4-dpdk-open/install/lib/python3.8/site-packages/IPython/core/history.py:226: UserWarning: IPython History requires SQLite, your history will not be saved
warn("IPython History requires SQLite, your history will not be saved")
Python 3.8.10 (3d8993a744:3d8993a744, May 15 2022, 21:53:16)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.18.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: bfrt.sample_lpm.enable
------> bfrt.sample_lpm.enable()

In [2]: pipe = bfrt.sample_lpm.pipe.MainControlImpl

In [3]: from netaddr import IPAddress

In [4]: pipe.ipv4_da_lpm.add_with_next_hop(dstAddr=IPAddress('192.168.2.0'), dstAddr_p_length=24,vport=2)
MatchSpec:
priority = 8
num_valid_match_bits:32
num_match_bytes:4
match_value_bits: 0xc0a80200
match_mask_bits: 0xffffff00
ActionSpec:
action_bmap:01
num_valid_action_data_bits:32
num_action_data_bytes:4
action_data: 0x00000002

In [5]: pipe.ipv4_da_lpm.get(dstAddr=IPAddress('192.168.2.0'),dstAddr_p_length=24)
Entry 0:
Entry key:
hdr.ipv4.dstAddr : ('0xC0A80200', '0x00000018')
Entry data (action : MainControlImpl.next_hop):
vport : 0x00000002

Out[5]: Entry for pipe.MainControlImpl.ipv4_da_lpm table.

In [6]: pipe.ipv4_da_lpm.dump
------> pipe.ipv4_da_lpm.dump()
----- ipv4_da_lpm Dump Start -----
Entry 0:
Entry key:
hdr.ipv4.dstAddr : ('0xC0A80200', '0x00000018')
Entry data (action : MainControlImpl.next_hop):
vport : 0x00000002

----- ipv4_da_lpm Dump End -----

In [7]: pipe.ipv4_da_lpm.get(dstAddr=IPAddress('192.168.2.0'),dstAddr_p_length=24)
Entry 0:
Entry key:
hdr.ipv4.dstAddr : ('0xC0A80200', '0x00000018')
Entry data (action : MainControlImpl.next_hop):
vport : 0x00000002

Out[7]: Entry for pipe.MainControlImpl.ipv4_da_lpm table.

In [8]: pipe.ipv4_da_lpm.add_with_next_hop(dstAddr=IPAddress('192.168.3.0'), dstAddr_p_length=24,vport=1)
MatchSpec:
priority = 8
num_valid_match_bits:32
num_match_bytes:4
match_value_bits: 0xc0a80300
match_mask_bits: 0xffffff00
ActionSpec:
action_bmap:01
num_valid_action_data_bits:32
num_action_data_bytes:4
action_data: 0x00000001

In [9]: pipe.ipv4_da_lpm.get(dstAddr=IPAddress('192.168.3.0'),dstAddr_p_length=24)
Entry 0:
Entry key:
hdr.ipv4.dstAddr : ('0xC0A80300', '0x00000018')
Entry data (action : MainControlImpl.next_hop):
vport : 0x00000001

Out[9]: Entry for pipe.MainControlImpl.ipv4_da_lpm table.

In [10]: pipe.ipv4_da_lpm.dump
-------> pipe.ipv4_da_lpm.dump()
----- ipv4_da_lpm Dump Start -----
Entry 0:
Entry key:
hdr.ipv4.dstAddr : ('0xC0A80200', '0x00000018')
Entry data (action : MainControlImpl.next_hop):
vport : 0x00000002

Entry 1:
Entry key:
hdr.ipv4.dstAddr : ('0xC0A80300', '0x00000018')
Entry data (action : MainControlImpl.next_hop):
vport : 0x00000001

----- ipv4_da_lpm Dump End -----

In [11]: pipe.ipv4_da_lpm.delete(dstAddr=IPAddress('192.168.2.0'),dstAddr_p_length=24)
MatchSpec:
priority = 8
num_valid_match_bits:32
num_match_bytes:4
match_value_bits: 0xc0a80200
match_mask_bits: 0xffffff00

In [12]: pipe.ipv4_da_lpm.delete(dstAddr=IPAddress('192.168.3.0'),dstAddr_p_length=24)
MatchSpec:
priority = 8
num_valid_match_bits:32
num_match_bytes:4
match_value_bits: 0xc0a80300
match_mask_bits: 0xffffff00

In [13]: pipe.ipv4_da_lpm.dump
-------> pipe.ipv4_da_lpm.dump()
----- ipv4_da_lpm Dump Start -----
Table pipe.MainControlImpl.ipv4_da_lpm has no entries.
----- ipv4_da_lpm Dump End -----

**Comment by riyazahm22** [comment]
PSA Test Example;
[root@localhost bin]# ./bf_switchd --install-dir $SDE_INSTALL --conf-file $SDE_INSTALL/share/examples/simple_l2l3_lpm/simple_l2l3_lpm.conf --init-mode=cold --status-port 7777 --skip-hld mkt
Skip-hld-mgrs list is mkt
Unknown skip-hld option m
Unknown skip-hld option k
Unknown skip-hld option t
Install dir: /home/rahmad/p4-dpdk-open/install (0x12deb80)
bf_switchd: system services initialized
bf_switchd: loading conf_file /home/rahmad/p4-dpdk-open/install/share/examples/simple_l2l3_lpm/simple_l2l3_lpm.conf...
bf_switchd: processing device configuration...
Configuration for dev_id 0
Family : dpdk
bf_switchd: processing P4 configuration...
P4 profile for dev_id 0
P4 EAL args dummy -n 4 -c 3
Debug CLI disable
num mempool objs 1
mempool_name: MEMPOOL0
buffer_size: 2304
pool_size: 1024
cache_size: 256
numa_node: 0
num P4 programs 1
p4_name: simple_l2l3_lpm
bfrt_config: /home/rahmad/p4-dpdk-open/install/share/examples/simple_l2l3_lpm/bf-rt.json
port_config: /home/rahmad/p4-dpdk-open/install/share/examples/simple_l2l3_lpm/port_config.json
p4_pipeline_name: ip
core_id: 1
numa_node: 0
context: /home/rahmad/p4-dpdk-open/install/share/examples/simple_l2l3_lpm/pipe/context.json
config: /home/rahmad/p4-dpdk-open/install/share/examples/simple_l2l3_lpm/pipe/simple_l2l3_lpm.spec
Pipes in scope [0 1 2 3 ]
Mirror Config
n_slots: 4
n_sessions: 64
fast_clone: 0
diag:
accton diag:
non_default_port_ppgs: 0
SAI default initialize: 1
Operational mode set to DPDK MODEL
Initialized the device types using PAL handler registration
STUB:port_mgr_platform_init DPDK
bf_switchd: drivers initialized
EAL: Detected CPU lcores: 16
EAL: Detected NUMA nodes: 1
EAL: Detected shared linkage of DPDK
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'PA'
EAL: Probe PCI driver: net_virtio (1af4:1000) device: 0000:00:03.0 (socket 0)
eth_virtio_pci_init(): Failed to init PCI device
EAL: Requested device 0000:00:03.0 cannot be used
EAL: Probe PCI driver: net_virtio (1af4:1000) device: 0000:00:04.0 (socket 0)
eth_virtio_pci_init(): Failed to init PCI device
EAL: Requested device 0000:00:04.0 cannot be used
TELEMETRY: No legacy callbacks, legacy socket not created

bf_switchd: dev_id 0 initialized

bf_switchd: initialized 1 devices
bf_switchd_switchsai_lib_init
starting SAI
bf_switchd: spawning cli server thread
bf_switchd: spawning driver shell
bf_switchd: server started - listening on port 9999

    ********************************************
    *      WARNING: Authorised Access Only     *
    ********************************************
bfshell> bfrt_python
cwd : /home/rahmad/p4-dpdk-open/install/bin

Devices found : [0]
We've found 1 p4 programs for device 0:
simple_l2l3_lpm
Creating tree for dev 0 and program simple_l2l3_lpm

/home/rahmad/p4-dpdk-open/install/lib/python3.8/site-packages/IPython/core/history.py:226: UserWarning: IPython History requires SQLite, your history will not be saved
warn("IPython History requires SQLite, your history will not be saved")
Python 3.8.10 (3d8993a744:3d8993a744, May 15 2022, 21:53:16)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.18.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: bfrt.simple_l2l3_lpm.enable
------> bfrt.simple_l2l3_lpm.enable()

In [2]: pipe = bfrt.simple_l2l3_lpm.ip.ingress

In [3]: from netaddr import IPAddress

In [4]: pipe.ipv4_lpm.add_with_send(dst_addr=IPAddress('192.168.2.0'),dst_addr_p_length=24, port=1)
MatchSpec:
priority = 8
num_valid_match_bits:32
num_match_bytes:4
match_value_bits: 0xc0a80200
match_mask_bits: 0xffffff00
ActionSpec:
action_bmap:01
num_valid_action_data_bits:32
num_action_data_bytes:4
action_data: 0x00000001

In [5]: pipe.ipv4_lpm.get(dst_addr=IPAddress('192.168.2.0'),dst_addr_p_length=24)
Entry 0:
Entry key:
hdr.ipv4.dst_addr : ('0xC0A80200', '0x00000018')
Entry data (action : ingress.send):
port : 0x00000001

Out[5]: Entry for ip.ingress.ipv4_lpm table.

In [6]: pipe.ipv4_lpm.dump
------> pipe.ipv4_lpm.dump()
----- ipv4_lpm Dump Start -----
Entry 0:
Entry key:
hdr.ipv4.dst_addr : ('0xC0A80200', '0x00000018')
Entry data (action : ingress.send):
port : 0x00000001

----- ipv4_lpm Dump End -----

In [7]: pipe.ipv4_lpm.delete(dst_addr=IPAddress('192.168.2.0'),dst_addr_p_length=24)
MatchSpec:
priority = 8
num_valid_match_bits:32
num_match_bytes:4
match_value_bits: 0xc0a80200
match_mask_bits: 0xffffff00

In [8]: pipe.ipv4_lpm.get(dst_addr=IPAddress('192.168.2.0'),dst_addr_p_length=24)
Error: table_entry_get failed on table ip.ingress.ipv4_lpm. [Object not found]
Out[8]: -1

In [9]: pipe.ipv4_lpm.add_with_send(dst_addr=IPAddress('192.168.2.0'),dst_addr_p_length=24, port=1)
MatchSpec:
priority = 8
num_valid_match_bits:32
num_match_bytes:4
match_value_bits: 0xc0a80200
match_mask_bits: 0xffffff00
ActionSpec:
action_bmap:01
num_valid_action_data_bits:32
num_action_data_bytes:4
action_data: 0x00000001

In [10]: pipe.ipv4_lpm.dump
-------> pipe.ipv4_lpm.dump()
----- ipv4_lpm Dump Start -----
Entry 0:
Entry key:
hdr.ipv4.dst_addr : ('0xC0A80200', '0x00000018')
Entry data (action : ingress.send):
port : 0x00000001

----- ipv4_lpm Dump End -----

In [11]: pipe.ipv4_lpm.delete(dst_addr=IPAddress('192.168.2.0'),dst_addr_p_length=24)
MatchSpec:
priority = 8
num_valid_match_bits:32
num_match_bytes:4
match_value_bits: 0xc0a80200
match_mask_bits: 0xffffff00

In [12]: pipe.ipv4_lpm.add_with_send(dst_addr=IPAddress('192.168.3.0'),dst_addr_p_length=24, port=1)
MatchSpec:
priority = 8
num_valid_match_bits:32
num_match_bytes:4
match_value_bits: 0xc0a80300
match_mask_bits: 0xffffff00
ActionSpec:
action_bmap:01
num_valid_action_data_bits:32
num_action_data_bytes:4
action_data: 0x00000001

In [13]: pipe.ipv4_lpm.add_with_send(dst_addr=IPAddress('192.168.4.0'),dst_addr_p_length=24, port=1)
MatchSpec:
priority = 8
num_valid_match_bits:32
num_match_bytes:4
match_value_bits: 0xc0a80400
match_mask_bits: 0xffffff00
ActionSpec:
action_bmap:01
num_valid_action_data_bits:32
num_action_data_bytes:4
action_data: 0x00000001

In [14]: pipe.ipv4_lpm.dump
-------> pipe.ipv4_lpm.dump()
----- ipv4_lpm Dump Start -----
Entry 0:
Entry key:
hdr.ipv4.dst_addr : ('0xC0A80300', '0x00000018')
Entry data (action : ingress.send):
port : 0x00000001

Entry 1:
Entry key:
hdr.ipv4.dst_addr : ('0xC0A80400', '0x00000018')
Entry data (action : ingress.send):
port : 0x00000001

----- ipv4_lpm Dump End -----

In [15]: pipe.ipv4_lpm.delete(dst_addr=IPAddress('192.168.4.0'),dst_addr_p_length=24)
MatchSpec:
priority = 8
num_valid_match_bits:32
num_match_bytes:4
match_value_bits: 0xc0a80400
match_mask_bits: 0xffffff00

In [16]: pipe.ipv4_lpm.dump
-------> pipe.ipv4_lpm.dump()
----- ipv4_lpm Dump Start -----
Entry 0:
Entry key:
hdr.ipv4.dst_addr : ('0xC0A80300', '0x00000018')
Entry data (action : ingress.send):
port : 0x00000001

----- ipv4_lpm Dump End -----

In [17]: pipe.ipv4_lpm.delete(dst_addr=IPAddress('192.168.3.0'),dst_addr_p_length=24)
MatchSpec:
priority = 8
num_valid_match_bits:32
num_match_bytes:4
match_value_bits: 0xc0a80300
match_mask_bits: 0xffffff00

In [18]: pipe.ipv4_lpm.dump
-------> pipe.ipv4_lpm.dump()
----- ipv4_lpm Dump Start -----
Table ip.ingress.ipv4_lpm has no entries.
----- ipv4_lpm Dump End -----

File: README [hunk 1]
```diff
@@ -44,7 +44,7 @@ p4-dpdk-target works in conjunction with following repos:
         in the install directory.
 
     P4-OVS:
-        (https://github.com/ipdk-io/ovs)
```
Comment (swaroopsarma) [inline README:47]: this is right. Its not p4-ovs

File: src/lld/dpdk/Makefile.am [hunk 2]
```diff
@@ -2,7 +2,7 @@ all:
 	$(shell ./apply_patch.sh > /dev/null 2>&1)
 	stat dpdk_src/build > /dev/null 2>&1 || \
 	(cd dpdk_src && meson -Dprefix=$(prefix) build && \
-	cd build && ninja -j48 && ninja install)
+	cd build && ninja && ninja install)
```
Comment (swaroopsarma) [inline src/lld/dpdk/Makefile.am:5]: just say ninja -j ; without number else the builds are very slow

Comment (riyazahm22) [reply src/lld/dpdk/Makefile.am:5]: it looks like only ninja -j is not working, it expects an argument.
  "ninja: option requires an argument -- 'j'"

File: README (hunk 1 above)
Comment (riyazahm22) [reply README:47]: addressed this commit.

**Review by jamescchoi — COMMENTED** [review]
I do have questions about why some of the things were done, but it's not really related to this review.  Whatever issues I see are more design issues, so I'm OK committing this code.

**Review by swaroopsarma — APPROVED** [review]
looks good to me

**Pull request merged**